### PR TITLE
Refactoring metric point storage

### LINKED ISSUE #1488: live ping metrics
Hi,

i know this site https://status.discordapp.com/ doesen't use cachet but it uses status.io or something like that and i was wondering is it possible to get a live ping stats thingy as that site but on cachet? so people can see actual ping ms and earlier stats??

### PR DESCRIPTION
**Pull request by jbrooksuk**
Closes #1488

---

Something I started yesterday but needed to get the branch up.

- [x] Add new `threshold` field to `metrics`.
- [x] Add new `counter` field to `metric_points`.
- [x] Update metric command handler to buffer within the threshold field.
- [x] Add a `calculated_value` result to the `MetricPointPresenter`.
- [x] Update MySQL repository to calculate points using the counter.
- [x] Update SQLite repository to calculate points using the counter.
- [x] Update Postgres repository to calculate points using the counter.

### CODE REVIEW

**Comment by jbrooksuk** [comment]
Maybe one for @joecohens to finish off if it needs doing before I'm back from my trip :)

**Comment by joecohens** [comment]
I'll finish this tomorrow :wink:

**Comment by jbrooksuk** [comment]
@joecohens it's just the metric repositories that need finishing now.

File: app/Bus/Handlers/Commands/Metric/AddMetricPointCommandHandler.php [hunk 1]
```diff
@@ -49,19 +51,37 @@ public function handle(AddMetricPointCommand $command)
...
+
+        event(new MetricPointWasAddedEvent($point));
+
+        return $point;
+    }
 
-        $metricPoint = MetricPoint::create($data);
+    protected function findOrCreatePoint(AddMetricPointCommand $command)
+    {
+        $buffer = Carbon::now()->subMinutes(5);
```
Comment (jbrooksuk) [inline app/Bus/Handlers/Commands/Metric/AddMetricPointCommandHandler.php]: Maybe we should make the buffer configurable per-metric?

Comment (jbrooksuk) [inline app/Bus/Handlers/Commands/Metric/AddMetricPointCommandHandler.php]: Ping @joecohens

Comment (joecohens) [inline app/Bus/Handlers/Commands/Metric/AddMetricPointCommandHandler.php]: :+1: Making this now

**Comment by joecohens** [comment]
Rebased

**Comment by jbrooksuk** [comment]
Just the repository stuff now and then we can merge.

File: app/Bus/Handlers/Commands/Metric/AddMetricPointCommandHandler.php [hunk 2]
```diff
@@ -65,23 +72,23 @@ public function handle(AddMetricPointCommand $command)
 
     protected function findOrCreatePoint(AddMetricPointCommand $command)
     {
-        $buffer = Carbon::now()->subMinutes(5);
+        $buffer = Carbon::now()->subMinutes($this->config->get('setting.metric_threshold', 5));
```
Comment (jbrooksuk) [inline app/Bus/Handlers/Commands/Metric/AddMetricPointCommandHandler.php]: I'm leaning towards thinking that this should be a per-metric setting?

Comment (GrahamCampbell) [inline app/Bus/Handlers/Commands/Metric/AddMetricPointCommandHandler.php]: We still shouldn't be injecting the config repo into handlers anyways. :P

Comment (joecohens) [inline app/Bus/Handlers/Commands/Metric/AddMetricPointCommandHandler.php]: @jbrooksuk Agreed, per metric should be better. @GrahamCampbell We are removing the config :stuck_out_tongue:

**Comment by jbrooksuk** [comment]
@joecohens just the last two repositories!

**Comment by GrahamCampbell** [comment]
NB We'll need to make a new 2.2 branch before 3.x features get merged.

**Comment by jbrooksuk** [comment]
👍

**Comment by jbrooksuk** [comment]
This is all tested and ready to be merged.

File: resources/views/dashboard/settings/app-setup.blade.php [hunk 3]
```diff
@@ -50,7 +50,7 @@
                                         <input type="number" max="100" name="app_incident_days" class="form-control" value="{{ Config::get('setting.app_incident_days', 7) }}">
                                     </div>
                                 </div>
-                            </div>
+                            </div>
```
Comment (GrahamCampbell) [inline resources/views/dashboard/settings/app-setup.blade.php]: ?

Comment (jbrooksuk) [inline resources/views/dashboard/settings/app-setup.blade.php]: I've no idea.

**Pull request merged**